modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientSql.java

```diff
@@ -242,7 +243,17 @@ public <T> CompletableFuture<AsyncResultSet<T>> executeAsync(
         if (transaction != null) {
             try {
                 //noinspection resource
-                return ClientTransaction.get(transaction).channel().serviceAsync(ClientOp.SQL_EXEC, payloadWriter, payloadReader);
+                return ClientLazyTransaction.ensureStarted(transaction, ch, null)
+                        .thenCompose(tx -> tx.channel().serviceAsync(ClientOp.SQL_EXEC, payloadWriter, payloadReader))
+                        .exceptionally(e -> {
+                            Throwable ex = unwrapCause(e);
+                            if (ex instanceof TransactionException) {
+                                var te = (TransactionException) ex;
+                                throw new SqlException(te.traceId(), te.code(), te.getMessage(), te);
```

Comment:
Here you create sql exception with txn error code, which is incorrect.
I don't think any unwapping is needed on client side. It's already done on server.
Same for line 258

Reply:
This was added specifically to match the embedded behavior and fix `ItSqlApiBaseTest#checkTransactionsWithDml` (this test runs against client and embedded APIs to ensure consistency):

https://github.com/apache/ignite-3/blob/0f9f1df95e0006edd9c0b0595a7c2082f2d5cb29/modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java#L249


Server-side mapping:

https://github.com/apache/ignite-3/blob/ee00a7c028ef6c8c1d8cdab0169fdd36deb1c5fe/modules/sql-engine/src/main/java/org/apache/ignite/internal/lang/SqlExceptionMapperUtil.java#L61

Reply:
> It's already done on server.

`Transaction is already finished` exception is thrown by the client in this case:

https://github.com/apache/ignite-3/blob/a70aef41a3dc56da9437acead50e76d17259967f/modules/client/src/main/java/org/apache/ignite/internal/client/tx/ClientTransaction.java#L167